openslides_backend/restrictions/restrictions.py

```diff
@@ -0,0 +1,148 @@
+from typing import Any, Callable, Dict, Type
+
+import fastjsonschema  # type: ignore
+from fastjsonschema import JsonSchemaException  # type: ignore
+
+from ..shared.exceptions import RestrictionException
+from ..shared.interfaces import LoggingModule, Services
+from ..shared.patterns import KEYSEPARATOR, Collection, FullQualifiedField
+from ..shared.schema import schema_version
+from .restrictions_interface import Payload, RestrictionResponse
+
+Restriction = Any  # TODO
+
+
+def prepare_restrictions_map() -> None:
+    """
+    This function just imports all restrictions modules so that the restrictions
+    are recognized by the system and the register decorator can do its work.
+
+    New modules have to be added here.
+    """
+    # from . import meeting, topic  # type: ignore # noqa
+    pass
+
+
+restrictions_map: Dict[Collection, Type[Restriction]] = {}
+
+
+def register_restriction(
+    name: Collection,
+) -> Callable[[Type[Restriction]], Type[Restriction]]:
+    """
+    Decorator to be used for restriction classes. Registers the class so that it
+    can be found by the handler.
+    """
+
+    def wrapper(restriction: Type[Restriction]) -> Type[Restriction]:
+        restrictions_map[name] = restriction
+        return restriction
+
+    return wrapper
+
+
+prepare_restrictions_map()
+
+
+payload_schema = fastjsonschema.compile(
+    {
+        "$schema": schema_version,
+        "title": "Schema for restrictions API",
+        "description": "An array of restriction blobs, i. e. bundles of user_id and full qualified fields.",
+        "type": "array",
```

Comment:
The autoupdate-service creates one request for one user. So the current code does not need the restricter to accept an array, it would only the the first element in the array.

It would be a challenge to bundle all requests from all connected users from one event. I understand, that there are performance improvements on the backend, when the requests are bundled. But I don't think this is possible with the current design of the autoupdate service.

Do you think, you can get the same performance improvements by using some sort of caching?

Do you think it is a big performance problem, when the autoupdate-service sends one request per user? Should I redesign the autoupdate service one request for all users are possible?

Reply:
I do not know if it is a big performance loss if you have to send me an array even if it always contains only one element.

Some kind of caching might be difficult, at least for the first draft we should deliver until March 1st or April 1st. I have no idea at the moment. For the first draft we just might scale up with more workers (it would be again the network stack that might be the bottle neck).

So I propose that the backend accepts arrays and the autoupdate service later decides whether to bundle or not. (And we even can later decide to remove the array in my code. This should be easier than the other way.)

Reply:
I don't think that it is any problem if I always send an array with one element.

I am also thinking about an optional caching-throtteling in the autoupdate-service. So an request to the restricter is not send immediately but the cache waits for some time (for example 10ms). All other requests in this timeframe are send together. You are right, that this is not necessary for the frist draft.

What do you think about not using an array but an object where the key is the user_id and the value the list of keys?
```
{
  5: ["motion/1/title", "user/5/name"],
  2: ["motion/1/title", "agenda/2/title"],
}
```
I think it is easier this way than keeping track of an index of an array.

The problem with the array is, that when (for any reason) one element of the response get lost, than the wrong connection gets the values for a different user.

Reply:
New format would be ok for me.

Reply:
If it is not urgent for you, then let me think a bit more about it. The disadvantage of the new format is, that when a user has many open connections, the autoupdate service has to distribute the keys of one user to the correct connections. With the array, I could skip this step.